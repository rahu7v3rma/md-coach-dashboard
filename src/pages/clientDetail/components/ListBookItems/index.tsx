import moment from 'moment';
import { FC, useState } from 'react';

import { logType } from '../../../../types/log';
import { logTypeDetails } from '../../../../utils/constants';
import { LogBookItem } from '../../components';
import {
    CardLogTitle,
    LessionView,
    LessonImage,
    LogBookItemContainer,
    MedicationDoseText,
    SubTitle,
    SubTitleText,
    SubView,
    TitleText,
    TitleView
} from '../../styles';

type Props = {
    list: logType[];
    groupTitle: string;
};

const roundNumbers = (item: any) => {
    const itemLocal = structuredClone(item);
    for (let key in itemLocal) {
        const itemNumberValue = Number(itemLocal[key]);
        if (itemNumberValue) {
            if (!Number.isInteger(itemNumberValue)) {
                itemLocal[key] = itemNumberValue.toFixed(2);
            }
        }
    }
    return itemLocal;
};

const minutesToHoursAndMinutes = (minutes: number): string => {
    const formattedHours = String(Math.floor(minutes / 60)).padStart(2, '0');
    const formattedMinutes = String(minutes % 60).padStart(2, '0');
    return `${formattedHours}:${formattedMinutes}`;
};

const ListBookItems: FC<Props> = ({ list, groupTitle }: Props) => {
    const [hasError, setHasError] = useState(false);

    function handleImageError() {
        setHasError(true);
    }

    return (
        <>
            <CardLogTitle>
                {moment(groupTitle, 'DD MMM YYYY').isSame(new Date(), 'day')
                    ? 'Today'
                    : groupTitle}
            </CardLogTitle>
            {list.map((item: any) => {
                const keyTyped = item.type as keyof typeof logTypeDetails;
                const icon: string = logTypeDetails[keyTyped].icon;
                const title: string = logTypeDetails[keyTyped].title;

                item = roundNumbers(item);

                return (
                    <LogBookItemContainer key={`log-${keyTyped}-${item.id}`}>
                        <LogBookItem
                            icon={icon}
                            title={title}
                            time={moment(
                                item.log_time,
                                'YYYY-MM-DD HH:mm:s'
                            ).format('HH:mm')}
                            image={item?.image}
                        >
                            <TitleView>
                                {item.duration_minutes && (
                                    <>
                                        <TitleText>
                                            {minutesToHoursAndMinutes(
                                                item.duration_minutes
                                            )}
                                        </TitleText>
                                    </>
                                )}
                                {item.drug_name && (
                                    <TitleText>
                                        {item.drug_name}
                                        {item.dose && (
                                            <MedicationDoseText>
                                                {`${item.amount} ${item.dose}`}
                                            </MedicationDoseText>
                                        )}
                                    </TitleText>
                                )}
                                {item.units && (
                                    <TitleText>{item.units}</TitleText>
                                )}
                                {item.injection_type && (
                                    <SubTitleText>
                                        {item.injection_type}
                                    </SubTitleText>
                                )}
                                {item.type !== 'UserMedication' && (
                                    <>
                                        {item.amount && (
                                            <TitleText>{item.amount}</TitleText>
                                        )}
                                        {!!item.measurement_type ? (
                                            <SubView>
                                                {item.unit && (
                                                    <SubTitle>
                                                        {item.unit}
                                                    </SubTitle>
                                                )}
                                                {item.measurement_type && (
                                                    <SubTitle>
                                                        {item.measurement_type}
                                                    </SubTitle>
                                                )}
                                            </SubView>
                                        ) : (
                                            item.unit && (
                                                <SubTitleText>
                                                    {item.unit}
                                                </SubTitleText>
                                            )
                                        )}
                                    </>
                                )}
                                {item.intensity && item.activity_type && (
                                    <SubView>
                                        <SubTitle>{item.intensity}</SubTitle>
                                        <SubTitle>
                                            {item.activity_type}
                                        </SubTitle>
                                    </SubView>
                                )}
                                {item.type === 'UserLesson' && (
                                    <LessionView>
                                        {!hasError && (
                                            <LessonImage
                                                src={item.lesson.icon}
                                                onError={handleImageError}
                                            />
                                        )}
                                        <SubTitle>{item.lesson.title}</SubTitle>
                                    </LessionView>
                                )}
                            </TitleView>
                        </LogBookItem>
                    </LogBookItemContainer>
                );
            })}
        </>
    );
};

export default ListBookItems;
