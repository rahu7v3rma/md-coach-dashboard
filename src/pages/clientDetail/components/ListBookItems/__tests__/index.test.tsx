import {
    RenderResult,
    fireEvent,
    render,
    screen
} from '@testing-library/react';
import moment from 'moment';
import React from 'react';
import { act } from 'react-dom/test-utils';

import ListBookItems from '..';

jest.mock('src/navigation');
global.structuredClone = (a: any) => JSON.parse(JSON.stringify(a));

const props = {
    list: [
        {
            id: 1,
            type: `UserLesson`,
            log_time: '2023-10-`12` 00:00:00',
            duration_minutes: '120',
            activity_type: 'Running',
            intensity: '10',
            lesson: {
                icon: 'lessonIcon',
                title: 'lessonTitle'
            },
            units: 1.2345
        }
    ],
    groupTitle: moment().format('DD MMM YYYY')
};

let view!: RenderResult;
beforeEach(() => {
    view = render(<ListBookItems {...props} />);
});

test('Snapshot', () => {
    expect(view.asFragment()).toMatchSnapshot();
});

test('renders ListBookItems component with a list of log items', () => {
    const logBookItem = view.container.children[1].children[0];
    const header = logBookItem.children[0];
    const titleContainer = header.children[0];
    expect(titleContainer.children[1].textContent).toBe('Lesson Completed');
    const timeContainer = header.children[1];
    expect(timeContainer.children[0].textContent).toBe(
        moment(props.list[0].log_time, 'YYYY-MM-DD HH:mm:s').format('HH:mm')
    );
    const bodyTitleView = logBookItem.children[1].children[0];
    expect(bodyTitleView.children[0].textContent).toBe(
        String(
            parseFloat((Number(props.list[0].duration_minutes) / 60).toFixed(2))
        )
    );
    expect(bodyTitleView.children[1].textContent).toBe(
        props.list[0].units.toFixed(2)
    );
    const bodySubView = bodyTitleView.children[2];
    expect(bodySubView.children[0].textContent).toBe(props.list[0].intensity);
    expect(bodySubView.children[1].textContent).toBe(
        props.list[0].activity_type
    );
    const bodyLessonView = bodyTitleView.children[3];
    expect(bodyLessonView.children[0].getAttribute('src')).toBe(
        props.list[0].lesson.icon
    );
    expect(bodyLessonView.children[1].textContent).toBe(
        props.list[0].lesson.title
    );
});

test('renders ListBookItems component with missing data', () => {
    view.rerender(<ListBookItems list={[]} groupTitle="" />);
    expect(view.container.textContent).toBe('');
});

test('handles image loading error', () => {
    const { container } = view;
    let lessonImage = container.querySelector(
        `img[src="${props.list[0].lesson.icon}"]`
    );
    expect(lessonImage).toBeInTheDocument();
    act(() => fireEvent.error(lessonImage as HTMLElement));
    lessonImage = container.querySelector(
        `img[src="${props.list[0].lesson.icon}"]`
    );
    expect(lessonImage).not.toBeInTheDocument();
});

test('handles decimal number formatting', () => {
    expect(
        screen.queryByText(props.list[0].units.toFixed(2))
    ).toBeInTheDocument();
});

test('renders ListBookItems component with "Today" as the group title', () => {
    const { container } = view;
    const groupTitle = container.firstElementChild?.textContent;
    expect(groupTitle).toBe('Today');
});
