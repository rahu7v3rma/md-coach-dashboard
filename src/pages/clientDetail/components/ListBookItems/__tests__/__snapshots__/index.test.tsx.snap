// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<DocumentFragment>
  <p
    class="sc-gTRrQi bFoqDV sc-jDnxRY sc-fGwHKD fPcaiw hIyJfw"
  >
    Today
  </p>
  <div
    class="sc-iBTzgk gVgZbe"
  >
    <div
      class="sc-jrrXlR ha-dviZ"
    >
      <div
        class="sc-IhgdQ gVMaCm"
      >
        <div
          class="sc-fxAUDG crypgK"
        >
          <img
            class="sc-guKdVK bNdiCm"
            src="[object Object]"
          />
          <p
            class="sc-gTRrQi bNPfnJ"
            color="#72996B"
            font-family="\\"Poppins\\", sans-serif"
            font-size="14"
            font-weight="600"
          >
            Lesson Completed
          </p>
        </div>
        <div
          class="sc-fxAUDG sc-dzIIgs crypgK"
        >
          <p
            class="sc-gTRrQi bKwhER"
            color="#271A51"
            font-family="\\"Poppins\\", sans-serif"
            font-size="14"
            font-weight="500"
          >
            00:00
          </p>
        </div>
      </div>
      <div
        class="sc-dGFpAT"
      >
        <div
          class="sc-iLBJAN gRTvBt"
        >
          <p
            class="sc-gTRrQi bFoqDV sc-irmnWS jgdkkg"
          >
            2
          </p>
          <p
            class="sc-gTRrQi bFoqDV sc-irmnWS jgdkkg"
          >
            1.23
          </p>
          <div
            class="sc-fYYabh RpqDi"
          >
            <p
              class="sc-gTRrQi bFoqDV sc-gKfoOY eFGkpe"
            >
              10
            </p>
            <p
              class="sc-gTRrQi bFoqDV sc-gKfoOY eFGkpe"
            >
              Running
            </p>
          </div>
          <div
            class="sc-jVOTke chnZQA"
          >
            <img
              class="sc-fnVmsx fsZnvi"
              src="lessonIcon"
            />
            <p
              class="sc-gTRrQi bFoqDV sc-gKfoOY eFGkpe"
            >
              lessonTitle
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
